### PR TITLE
Payload parameters

### PR DESCRIPTION
**Pull request by foxted**
I'm new to Terraform, but I gave it a shot to try and help. If you had a different implementation in mind, please let me know, happy to help 👍🏻 

Example with `repositories`: 

```terraform
module "gcr_cleaner" {
  source = "mirakl/gcr-cleaner/google"
  version = "0.5.0"

  app_engine_application_location = "northamerica-northeast1"

  gcr_repositories = [
    {
      repositories = [
        {
          repo = "gcr.io/example/repository"
          allow_tagged = true
          keep = 10
          grace = "730s"       
        }
      ]
    },
  ]
}
```

Example with `clean_all`:

```terraform
module "gcr_cleaner" {
  source = "mirakl/gcr-cleaner/google"
  version = "0.5.0"

  app_engine_application_location = "northamerica-northeast1"

  gcr_repositories = [
    {
      clean_all = true
      parameters = {
        allow_tagged = true
        keep = 10
        grace = "730s"
      }
    },
  ]
}
```

### CODE REVIEW

**Comment by anouarchattouna** [comment]
Hello @foxted, thank you for opening this PR.
Could you please update [examples](https://github.com/mirakl/terraform-google-gcr-cleaner/tree/main/examples) accordingly?

**Comment by foxted** [comment]
No worries, updated the examples, sprinkling example on how to use the payload parameters.

File: locals.tf [hunk 1]
```diff
@@ -21,14 +21,26 @@ locals {
   # for projects having explicit repositories definition
   repositories = flatten([
     for gcr in var.gcr_repositories : [
-      for repo in gcr.repositories : gcr.storage_region != null ? "${gcr.storage_region}.gcr.io/${gcr.project_id != null ? gcr.project_id : local.google_project_id}/${repo}" : "gcr.io/${gcr.project_id != null ? gcr.project_id : local.google_project_id}/${repo}"
+      for repo in gcr.repositories : gcr.storage_region != null
+        ? merge({
```
Comment (anouarchattouna) [inline locals.tf:25]:
```suggestion
      ? merge(
        repo,
        {
          repo         = "${gcr.storage_region}.gcr.io/${gcr.project_id != null ? gcr.project_id : local.google_project_id}/${repo.repo}"
          grace        = repo.grace != null ? repo.grace : "0"
          allow_tagged = repo.allow_tagged != null ? repo.allow_tagged : false
          keep         = repo.keep != null ? repo.keep : 0
          tag_filter   = repo.tag_filter != null ? repo.tag_filter : ""
        }
      )
      : merge(
        repo,
        {
          repo         = "gcr.io/${gcr.project_id != null ? gcr.project_id : local.google_project_id}/${repo.repo}"
          grace        = repo.grace != null ? repo.grace : "0"
          allow_tagged = repo.allow_tagged != null ? repo.allow_tagged : false
          keep         = repo.keep != null ? repo.keep : 0
          tag_filter   = repo.tag_filter != null ? repo.tag_filter : ""
        }
      )
```
* need to switch maps order in order to keep elements from the 2end map
* need to handle null values

File: locals.tf [hunk 2]
```diff
@@ -21,14 +21,26 @@ locals {
...
+          repo
+        )
     ] if gcr.repositories != null
   ])
 
   # merge all fetched repositories with external data in one list
   fetched_repositories = flatten([
     for data in data.external.this : [
-      for repo in jsondecode(data.result.repositories) : repo
+      for repo in jsondecode(data.result.repositories) : merge({ "repo": repo.name }, local.options[repo.project_id])
```
Comment (anouarchattouna) [inline locals.tf:39]:
```suggestion
      for repo in jsondecode(data.result.repositories) : merge(
        local.options[repo.project_id],
        {
          repo         = repo.name
          grace        = local.options[repo.project_id].grace != null ? local.options[repo.project_id].grace : "0"
          allow_tagged = local.options[repo.project_id].allow_tagged != null ? local.options[repo.project_id].allow_tagged : false
          keep         = local.options[repo.project_id].keep != null ? local.options[repo.project_id].keep : 0
          tag_filter   = local.options[repo.project_id].tag_filter != null ? local.options[repo.project_id].tag_filter : ""
        }
      )
```
* need to switch maps order in order to keep elements from the 2end map
* need to handle null values

File: main.tf [hunk 3]
```diff
@@ -54,18 +54,18 @@ resource "google_app_engine_application" "this" {
 # 2. Cloud Scheduler job that triggers an action via HTTP
 resource "google_cloud_scheduler_job" "this" {
   for_each = {
-    for repo in concat(local.fetched_repositories, local.repositories) : repo => repo
+    for idx, repo in concat(local.fetched_repositories, local.repositories): idx => repo
```
Comment (anouarchattouna) [inline main.tf:57]:
```suggestion
    for idx, repo in concat(local.fetched_repositories, local.repositories) : idx => repo
```

**Review by anouarchattouna — CHANGES_REQUESTED** [review]
Please apply those changes on your project and run the jobs for validation

**Comment by foxted** [comment]
@anouarchattouna I saw your comments, I'll update this tonight 👍🏻

**Comment by foxted** [comment]
Changes applied :)

File: locals.tf [hunk 4]
```diff
@@ -21,14 +21,45 @@ locals {
   # for projects having explicit repositories definition
   repositories = flatten([
     for gcr in var.gcr_repositories : [
-      for repo in gcr.repositories : gcr.storage_region != null ? "${gcr.storage_region}.gcr.io/${gcr.project_id != null ? gcr.project_id : local.google_project_id}/${repo}" : "gcr.io/${gcr.project_id != null ? gcr.project_id : local.google_project_id}/${repo}"
+      for repo in gcr.repositories : gcr.storage_region != null
+        ? merge(
+            repo,
+            {
+              repo = "${gcr.storage_region}.gcr.io/${gcr.project_id != null ? gcr.project_id : local.google_project_id}/${repo.repo}",
```
Comment (anouarchattouna) [inline locals.tf:28]:
```suggestion
              repo = "${gcr.storage_region}.gcr.io/${gcr.project_id != null ? gcr.project_id : local.google_project_id}/${repo.repo}"
```

File: locals.tf [hunk 5]
```diff
@@ -21,14 +21,45 @@ locals {
...
+              grace = repo.grace != null ? repo.grace : "0",
```
Comment (anouarchattouna) [inline locals.tf:29]:
```suggestion
              grace = repo.grace != null ? repo.grace : "0"
```

File: locals.tf [hunk 6]
```diff
@@ -21,14 +21,45 @@ locals {
...
+              allow_tagged = repo.allow_tagged != null ? repo.allow_tagged : false,
```
Comment (anouarchattouna) [inline locals.tf:30]:
```suggestion
              allow_tagged = repo.allow_tagged != null ? repo.allow_tagged : false
```

File: locals.tf [hunk 7]
```diff
@@ -21,14 +21,45 @@ locals {
...
+              tag_filter = repo.tag_filter != null ? repo.tag_filter : ""
+            }
+        )
+        : merge(
+            repo,
+            {
+              repo = "gcr.io/${gcr.project_id != null ? gcr.project_id : local.google_project_id}/${repo.repo}",
```
Comment (anouarchattouna) [inline locals.tf:37]:
```suggestion
              repo = "gcr.io/${gcr.project_id != null ? gcr.project_id : local.google_project_id}/${repo.repo}"
```

File: locals.tf [hunk 8]
```diff
@@ -21,14 +21,45 @@ locals {
...
+              grace = repo.grace != null ? repo.grace : "0",
```
Comment (anouarchattouna) [inline locals.tf:38]:
```suggestion
              grace = repo.grace != null ? repo.grace : "0"
```

File: locals.tf [hunk 9]
```diff
@@ -21,14 +21,45 @@ locals {
...
+              allow_tagged = repo.allow_tagged != null ? repo.allow_tagged : false,
```
Comment (anouarchattouna) [inline locals.tf:39]:
```suggestion
              allow_tagged = repo.allow_tagged != null ? repo.allow_tagged : false
```

File: locals.tf [hunk 10]
```diff
@@ -21,14 +21,45 @@ locals {
...
 
   # merge all fetched repositories with external data in one list
   fetched_repositories = flatten([
     for data in data.external.this : [
-      for repo in jsondecode(data.result.repositories) : repo
+      for repo in jsondecode(data.result.repositories) : merge(
+        local.options[repo.project_id],
+        {
+          "repo": repo.name
+          grace = local.options[repo.project_id].grace != null ? local.options[repo.project_id].grace : "0",
```
Comment (anouarchattouna) [inline locals.tf:53]:
```suggestion
          grace = local.options[repo.project_id].grace != null ? local.options[repo.project_id].grace : "0"
```

File: locals.tf [hunk 11]
```diff
@@ -21,14 +21,45 @@ locals {
...
+          allow_tagged = local.options[repo.project_id].allow_tagged != null ? local.options[repo.project_id].allow_tagged : false,
```
Comment (anouarchattouna) [inline locals.tf:54]:
```suggestion
          allow_tagged = local.options[repo.project_id].allow_tagged != null ? local.options[repo.project_id].allow_tagged : false
```

File: locals.tf [hunk 12]
```diff
@@ -21,14 +21,45 @@ locals {
...
+          keep = local.options[repo.project_id].keep != null ? local.options[repo.project_id].keep : 0,
```
Comment (anouarchattouna) [inline locals.tf:55]:
```suggestion
          keep = local.options[repo.project_id].keep != null ? local.options[repo.project_id].keep : 0
```

**Review by anouarchattouna — CHANGES_REQUESTED** [review]

**Comment by CrackerJackMack** [comment]
Pulled the latest in main, using multiple lines in the same project creates a key conflict.  The use case (requirement) for this is that using allow_tagged and tag_filter, you need to run them separated in order for the total count of images to be unique per tag.  Using a collapsed regex `^(main|master|dev)` results in only 10 images out a list containing all of those tags.  This produces less than ideal results.

```
module "gcr_cleaner" {
  source = "./gcr"

  gcr_repositories = [
    {
      // delete all untagged images
      project_id = "fart"
      clean_all  = true
    },
    {
      // in all repositories, keep 10 `main` builds, ignore anything newer than 5 days
      project_id = "fart"
      clean_all  = true
      parameters = {
        allow_tagged = true
        keep         = 10
        grace        = "5d"
        tag_filter   = "^main.+$"
      }
    },
    {
      // in all repositories, keep 10 `dev` builds, ignore anything newer than 5 days
      project_id = "fart"
      clean_all  = true
      parameters = {
        allow_tagged = true
        keep         = 10
        grace        = "5d"
        tag_filter   = "^dev.+$"
      }
    },
    {
      // keep only 2 master builds if they existed, just in case.
      project_id = "fart"
      clean_all  = true
      parameters = {
        allow_tagged = true
        keep         = 2
        tag_filter   = "^master.+$"
      }
    }
  ]
}
```

generates 
```
│ Error: Duplicate object key
│
│   on gcr/locals.tf line 63, in locals:
│   62:   options = {
│   63:     for gcr in var.gcr_repositories : (gcr.project_id != null ? gcr.project_id : local.google_project_id) => gcr.parameters
│   64:   }
│     ├────────────────
│     │ gcr.project_id is "fart"
│     │ local.google_project_id is "fart"
│
│ Two different items produced the key "fart" in this 'for' expression. If duplicates are expected, use the ellipsis (...) after the value expression to enable grouping by key.
```

**Comment by anouarchattouna** [comment]
@CrackerJackMack thank you for that.
@foxted any thoughts? Could you please have a look on it?

**Comment by foxted** [comment]
We might need to randomize or serialize the project names in the for loop. I didn't consider the use case where multiple configurations would be used for the same project. I'll have to go back to the drawing board.

**Comment by anouarchattouna** [comment]
@CrackerJackMack, see #24

**Comment by anouarchattouna** [comment]
Feature implemented in https://github.com/mirakl/terraform-google-gcr-cleaner/pull/24

**Pull request closed without merging**